### PR TITLE
KATI-70_Doctor-Working-Hour-Database

### PR DESCRIPTION
**Pull request by vtnlinh95**

### CODE REVIEW

File: com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java [hunk 1]
```diff
@@ -154,6 +150,30 @@ protected void insertToDoctorReferenceTables(DoctorUserEntity entity) throws SQL
...
 
+	protected void insertWorkingHour(DoctorUserEntity doctorUserEntity) throws SQLException {
+		List<OpeningHour> hours = doctorUserEntity.getFacility().get(0).getOpeningHours();
+		PreparedStatement stmt = null;
+		StringBuilder builder = new StringBuilder();
+
+		builder.append("INSERT INTO ");
+		builder.append(Doctor_FacilityColumn.TABLE_NAME);
+		builder.append(" VALUES (?, ?, ?, ?, ?);");
+		try{
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java]: re-format this file

File: com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java [hunk 2]
```diff
@@ -154,6 +150,30 @@ protected void insertToDoctorReferenceTables(DoctorUserEntity entity) throws SQL
...
+			for(int i=0; i<hours.size();++i){
+				stmt = con.prepareStatement(builder.toString());
+				stmt.setInt(1,Integer.parseInt(doctorUserEntity.getId()));
+				stmt.setInt(2,Integer.parseInt(doctorUserEntity.getFacility().get(0).getId()));
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java]: Extract doctorUserEntity.getId() and doctorUserEntity.getFacility().get(0).getId() for re-use with this loop

File: com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java [hunk 3]
```diff
@@ -154,6 +150,30 @@ protected void insertToDoctorReferenceTables(DoctorUserEntity entity) throws SQL
...
+				stmt.setInt(3,hours.get(i).getDayOfTheWeek().getCode());
+				stmt.setTime(4, hours.get(i).getOpenTime());
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java]: Extract hours.get(i) to re-use with this loop

File: com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java [hunk 4]
```diff
@@ -212,4 +232,112 @@ public DoctorUserEntity searchDoctor(UserEntity entity) throws ClassNotFoundExce
...
+					throw new Exception("Edit Working Hours failed");
+				}
+			}
+		}
+		finally{
+			stmt.close();
+		}
+	}
+
+	public List<OpeningHour> getWorkingHourAll(int doctorID) throws SQLException {
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java]: Method name should be getAllWorkingHourByDoctorId

**Comment by vtnlinh95** [comment]
@hieumai  fixed

File: com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java [hunk 5]
```diff
@@ -154,6 +151,32 @@ protected void insertToDoctorReferenceTables(DoctorUserEntity entity) throws SQL
...
+	protected void insertWorkingHour(DoctorUserEntity doctorUserEntity) throws SQLException {
+		List<OpeningHour> hours = doctorUserEntity.getFacility().get(0).getOpeningHours();
+		PreparedStatement stmt = null;
+		StringBuilder builder = new StringBuilder();
+		String doctorID = doctorUserEntity.getId();
+		String facilityID = doctorUserEntity.getFacility().get(0).getId();
+		builder.append("INSERT INTO ");
+		builder.append(Doctor_FacilityColumn.TABLE_NAME);
+		builder.append(" VALUES (?, ?, ?, ?, ?);");
+		try{
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java]: re-format this file

File: com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java [hunk 6]
```diff
@@ -154,6 +151,32 @@ protected void insertToDoctorReferenceTables(DoctorUserEntity entity) throws SQL
...
+			for(int i=0; i<hours.size();++i){
+				OpeningHour hour = hours.get(i);
+				stmt = con.prepareStatement(builder.toString());
+				stmt.setInt(1, Integer.parseInt(doctorID));
+				stmt.setInt(2, Integer.parseInt(facilityID));
+				stmt.setInt(3, hour.getDayOfTheWeek().getCode());
+				stmt.setTime(4, hour.getOpenTime());
+				stmt.setTime(5, hour.getCloseTime());
+				stmt.executeUpdate();
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java:172]: except from GET methods, all other methods that changing database need to rollback if exceptions happened

File: com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java [hunk 7]
```diff
@@ -212,4 +235,121 @@ public DoctorUserEntity searchDoctor(UserEntity entity) throws ClassNotFoundExce
...
+		try{
+			for (int i = 0; i<workingHours.size();++i){
+				stmt = con.prepareStatement(builder.toString());
+				stmt.setInt(1, workingHours.get(i).getDayOfTheWeek().getCode());
+				stmt.setTime(2, workingHours.get(i).getOpenTime());
+				stmt.setTime(3, workingHours.get(i).getCloseTime());
+				stmt.setInt(4, doctorID);
+				stmt.setInt(5, facilityID);
+				if(stmt.executeUpdate() == 0){
+					throw new Exception("Edit Working Hours failed");
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java:317]: same here too

File: com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java [hunk 8]
```diff
@@ -212,4 +235,121 @@ public DoctorUserEntity searchDoctor(UserEntity entity) throws ClassNotFoundExce
...
+				while(rs.next()){
+					allFacilityID.add(rs.getInt(Doctor_FacilityColumn.FACILITY_ID.getColumnName()));
+				}
+				return allFacilityID;
+			}
+		}
+		finally{
+			stmt.close();
+		}
+		return null;
```
Comment (hieumai) [inline com.kms.cura.dal/src/com/kms/cura/dal/database/DoctorUserDatabaseHelper.java]: Should return empty list instead of null

**Comment by vtnlinh95** [comment]
@hieumai fixed

**Comment by hieumai** [comment]
Looks OK to me. @huuduy216  and @toannguyen71195  please verify.

**Comment by toannguyen71195** [comment]
@hieumai Verified

**Comment by vtnlinh95** [comment]
@huuduy216  fix

**Pull request closed without merging**